packages/bus/src/brokers/RMQEventBus.ts

```diff
@@ -0,0 +1,479 @@
+  ): Promise<void> {
+    if (!this.channel) {
+      throw new IllegalOperation(this);
+    }
+
+    const handlerInstance = this.getHandlerInstance(type);
+
+    const eventNames: string[] = this.reflectEventsNames(type, handlerInstance);
+
+    await Promise.all(
+      eventNames.map(eventName =>
+        this.unregisterEventHandler(handlerInstance, eventName)
+      )
+    );
+  }
+
+  public async publish<T>(event: Event<T>): Promise<void> {
+    if (!this.channel) {
+      throw new IllegalOperation(this);
+    }
+
+    const { type, payload, correlationId, createdAt } = event;
+
+    await this.backOffRetry.acquire(() =>
+      this.channel?.publish(
+        this.busConfig.exchange,
+        type,
+        Buffer.from(JSON.stringify(payload)),
+        {
+          type,
+          correlationId,
+          contentType: 'application/json',
+          mandatory: true,
+          persistent: true,
+          replyTo: this.REPLY_QUEUE_NAME,
+          timestamp: createdAt.getTime()
+        }
+      )
+    );
+  }
+
+  public async execute<T, R>(command: Command<T, R>): Promise<R> {
+    const { correlationId, ttl, expectReply } = command;
+
+    let waiter = Promise.resolve<unknown>(undefined);
+
+    if (expectReply) {
+      process.nextTick(() => {
+        waiter = this.expectReply<R>(correlationId, ttl);
+      });
+    }
+
+    try {
+      this.sendCommandToQueue(command);
+
+      return (await waiter) as R;
+    } finally {
+      this.subject.removeAllListeners(correlationId);
+    }
+  }
+
+  public async destroy(): Promise<void> {
+    if (!this.client) {
+      throw new IllegalOperation(this);
+    }
+
+    if (this.channel) {
+      await this.channel.waitForConfirms();
+      await Promise.all(
+        this.consumerTags.map(consumerTag => this.channel?.cancel(consumerTag))
+      );
+      await this.channel.close();
+    }
+
+    await this.client.close();
+    this.clear();
+  }
+
+  private async registerEventHandler<T, R>(
+    handlerInstance: EventHandler<T, R>,
+    eventName: string
+  ) {
+    const eventHandlers = this.handlers.get(eventName);
+    if (Array.isArray(eventHandlers)) {
+      eventHandlers.push(handlerInstance);
+    } else {
+      this.handlers.set(eventName, [handlerInstance]);
+      await this.bindQueue(eventName);
+    }
+  }
+
+  private async bindQueue(eventName: string) {
+    if (!this.channel) {
+      throw new IllegalOperation(this);
+    }
+
+    await this.channel.bindQueue(
+      this.busConfig.clientQueue,
+      this.busConfig.exchange,
+      eventName
+    );
+  }
+
+  private async unregisterEventHandler<T, R>(
+    handlerInstance: EventHandler<T, R>,
+    eventName: string
+  ) {
+    const eventHandlers = this.handlers.get(eventName);
+    if (!eventHandlers) {
+      return;
+    }
+
+    const handlerIndex = eventHandlers.findIndex(
+      e =>
+        Object.getPrototypeOf(e).constructor.name ===
+        Object.getPrototypeOf(handlerInstance).constructor.name
+    );
```

Comment:
First of all, you have the dedicated function that does that same thing. Secondly, you should remove the concrete instance by link.

```suggestion
    const handlerIndex = eventHandlers.indexOf(handler);
```